# بسم الله الرحمن الرحيم

In [23]:
from utils import *
import pandas as pd 
from sklearn.model_selection import train_test_split


## Reading the Dataset, cleaning , preparing it for the training 

In [24]:
df = pd.read_csv("dataset/body_level_classification_train.csv")

In [25]:
df_original= cleaning_data(df).copy()

In [26]:
all_class_train,all_class_test=data_spliting(df_original,1/5)

In [27]:
oversampled= over_sampling(all_class_train,sampling_ratio=0.5)

In [28]:
X_train=oversampled.drop(["Body_Level"],axis=1).to_numpy()
Y_train=oversampled["Body_Level"].to_numpy()
X_test =all_class_test.drop(["Body_Level"],axis=1).to_numpy()
Y_test =all_class_test["Body_Level"].to_numpy()

In [29]:
X_train, scaler= data_scaling(X_train)

In [30]:
X_val, X_test, Y_val, Y_test = train_test_split(X_test
                                                     , Y_test, test_size=0.5, random_state=42)

In [31]:
X_val = scaler.transform(X_val)
X_test=scaler.transform(X_test)

In [32]:
# model=build_model(2,0.01)

model_one_layer  = keras.Sequential([keras.Input(shape=(23)),
                           layers.Dense(4,activation='relu'),
                           
                        layers.Dense(4,activation='softmax')])
model_one_layer.compile(loss=SparseCategoricalFocalLoss(from_logits=False,gamma=2),
          optimizer=keras.optimizers.Adam(learning_rate=0.01),
          metrics=["accuracy"])

In [33]:
model = model_one_layer
history = model.fit(X_train,Y_train,validation_data=[X_val,Y_val] 
             ,batch_size=32, epochs=400 , verbose =1, 
                       callbacks=[CustomCallback("NNModels/after_sampling/fifth_model.h5")])

Epoch 1/400
46/46 [==============================] - 1s 6ms/step - loss: 0.6826 - accuracy: 0.4884 - val_loss: 0.4064 - val_accuracy: 0.6122
Epoch 2/400
46/46 [==============================] - 0s 2ms/step - loss: 0.3486 - accuracy: 0.6612 - val_loss: 0.2492 - val_accuracy: 0.7619
Epoch 3/400
46/46 [==============================] - 0s 2ms/step - loss: 0.2508 - accuracy: 0.7789 - val_loss: 0.1900 - val_accuracy: 0.8027
Epoch 4/400
46/46 [==============================] - 0s 3ms/step - loss: 0.1916 - accuracy: 0.8354 - val_loss: 0.1431 - val_accuracy: 0.9116
Epoch 5/400
46/46 [==============================] - 0s 3ms/step - loss: 0.1460 - accuracy: 0.8864 - val_loss: 0.1054 - val_accuracy: 0.9048
Epoch 6/400
46/46 [==============================] - 0s 3ms/step - loss: 0.1132 - accuracy: 0.9095 - val_loss: 0.0823 - val_accuracy: 0.9388
Epoch 7/400
30/46 [==================>...........] - ETA: 0s - loss: 0.0962 - accuracy: 0.9260
Yes You are here 0.9238095283508301 0.9251700639724731
Kera

30/46 [==================>...........] - ETA: 0s - loss: 0.0295 - accuracy: 0.9740
Yes You are here 0.9727891087532043 0.9659863710403442
Keras weights file (<HDF5 file "variables.h5" (mode r+)>) saving:
...layers
......dense
.........vars
............0
............1
......dense_1
.........vars
............0
............1
...metrics
......mean
.........vars
............0
............1
......mean_metric_wrapper
.........vars
............0
............1
...optimizer
......vars
.........0
.........1
.........2
.........3
.........4
.........5
.........6
.........7
.........8
...vars
Keras model archive saving:
File Name                                             Modified             Size
variables.h5                                   2023-05-15 10:23:48        21392
config.json                                    2023-05-15 10:23:48         2011
metadata.json                                  2023-05-15 10:23:48           64
46/46 [==============================] - 0s 4ms/step - loss: 0.02

Epoch 31/400
46/46 [==============================] - 0s 3ms/step - loss: 0.0177 - accuracy: 0.9803 - val_loss: 0.0313 - val_accuracy: 0.9796
Epoch 32/400
46/46 [==============================] - 0s 2ms/step - loss: 0.0172 - accuracy: 0.9816 - val_loss: 0.0328 - val_accuracy: 0.9728
Epoch 33/400
46/46 [==============================] - 0s 2ms/step - loss: 0.0167 - accuracy: 0.9782 - val_loss: 0.0320 - val_accuracy: 0.9728
Epoch 34/400
24/46 [==============>...............] - ETA: 0s - loss: 0.0152 - accuracy: 0.9857
Yes You are here 0.980272114276886 0.9863945841789246
Keras weights file (<HDF5 file "variables.h5" (mode r+)>) saving:
...layers
......dense
.........vars
............0
............1
......dense_1
.........vars
............0
............1
...metrics
......mean
.........vars
............0
............1
......mean_metric_wrapper
.........vars
............0
............1
...optimizer
......vars
.........0
.........1
.........2
.........3
.........4
.........5
.........6
.....

46/46 [==============================] - 0s 3ms/step - loss: 0.0102 - accuracy: 0.9891 - val_loss: 0.0252 - val_accuracy: 0.9864
Epoch 50/400
46/46 [==============================] - 0s 2ms/step - loss: 0.0106 - accuracy: 0.9864 - val_loss: 0.0421 - val_accuracy: 0.9864
Epoch 51/400
30/46 [==================>...........] - ETA: 0s - loss: 0.0199 - accuracy: 0.9812
Yes You are here 0.9850339889526367 0.9931972622871399
Keras weights file (<HDF5 file "variables.h5" (mode r+)>) saving:
...layers
......dense
.........vars
............0
............1
......dense_1
.........vars
............0
............1
...metrics
......mean
.........vars
............0
............1
......mean_metric_wrapper
.........vars
............0
............1
...optimizer
......vars
.........0
.........1
.........2
.........3
.........4
.........5
.........6
.........7
.........8
...vars
Keras model archive saving:
File Name                                             Modified             Size
variables.h5         

46/46 [==============================] - 0s 2ms/step - loss: 0.0073 - accuracy: 0.9905 - val_loss: 0.0379 - val_accuracy: 0.9864
Epoch 88/400
46/46 [==============================] - 0s 2ms/step - loss: 0.0062 - accuracy: 0.9946 - val_loss: 0.0336 - val_accuracy: 0.9864
Epoch 89/400
46/46 [==============================] - 0s 3ms/step - loss: 0.0073 - accuracy: 0.9898 - val_loss: 0.0367 - val_accuracy: 0.9796
Epoch 90/400
46/46 [==============================] - 0s 3ms/step - loss: 0.0077 - accuracy: 0.9918 - val_loss: 0.0426 - val_accuracy: 0.9796
Epoch 91/400
46/46 [==============================] - 0s 2ms/step - loss: 0.0059 - accuracy: 0.9932 - val_loss: 0.0346 - val_accuracy: 0.9796
Epoch 92/400
46/46 [==============================] - 0s 2ms/step - loss: 0.0066 - accuracy: 0.9912 - val_loss: 0.0344 - val_accuracy: 0.9796
Epoch 93/400
46/46 [==============================] - 0s 2ms/step - loss: 0.0063 - accuracy: 0.9918 - val_loss: 0.0318 - val_accuracy: 0.9864
Epoch 94/400
46/46 

File Name                                             Modified             Size
variables.h5                                   2023-05-15 10:24:00        21392
config.json                                    2023-05-15 10:24:00         2011
metadata.json                                  2023-05-15 10:24:00           64
46/46 [==============================] - 0s 3ms/step - loss: 0.0073 - accuracy: 0.9912 - val_loss: 0.0311 - val_accuracy: 0.9932
Epoch 112/400
46/46 [==============================] - 0s 2ms/step - loss: 0.0051 - accuracy: 0.9939 - val_loss: 0.0322 - val_accuracy: 0.9864
Epoch 113/400
46/46 [==============================] - 0s 2ms/step - loss: 0.0043 - accuracy: 0.9952 - val_loss: 0.0376 - val_accuracy: 0.9660
Epoch 114/400
46/46 [==============================] - 0s 2ms/step - loss: 0.0144 - accuracy: 0.9857 - val_loss: 0.0361 - val_accuracy: 0.9864
Epoch 115/400
46/46 [==============================] - 0s 2ms/step - loss: 0.0087 - accuracy: 0.9898 - val_loss: 0.0432 - 

File Name                                             Modified             Size
variables.h5                                   2023-05-15 10:24:03        21392
config.json                                    2023-05-15 10:24:03         2011
metadata.json                                  2023-05-15 10:24:03           64
46/46 [==============================] - 0s 3ms/step - loss: 0.0027 - accuracy: 0.9980 - val_loss: 0.0332 - val_accuracy: 0.9932
Epoch 139/400
46/46 [==============================] - 0s 2ms/step - loss: 0.0031 - accuracy: 0.9966 - val_loss: 0.0394 - val_accuracy: 0.9864
Epoch 140/400
27/46 [================>.............] - ETA: 0s - loss: 0.0032 - accuracy: 0.9965    
Yes You are here 0.9972789287567139 0.9931972622871399
Keras weights file (<HDF5 file "variables.h5" (mode r+)>) saving:
...layers
......dense
.........vars
............0
............1
......dense_1
.........vars
............0
............1
...metrics
......mean
.........vars
............0
............1
..

Epoch 167/400
46/46 [==============================] - 0s 2ms/step - loss: 0.0094 - accuracy: 0.9898 - val_loss: 0.0551 - val_accuracy: 0.9728
Epoch 168/400
46/46 [==============================] - 0s 2ms/step - loss: 0.0044 - accuracy: 0.9959 - val_loss: 0.0590 - val_accuracy: 0.9796
Epoch 169/400
46/46 [==============================] - 0s 3ms/step - loss: 0.0037 - accuracy: 0.9959 - val_loss: 0.0641 - val_accuracy: 0.9864
Epoch 170/400
46/46 [==============================] - 0s 3ms/step - loss: 0.0027 - accuracy: 0.9966 - val_loss: 0.0536 - val_accuracy: 0.9864
Epoch 171/400
46/46 [==============================] - 0s 2ms/step - loss: 0.0028 - accuracy: 0.9959 - val_loss: 0.0596 - val_accuracy: 0.9864
Epoch 172/400
46/46 [==============================] - 0s 2ms/step - loss: 0.0026 - accuracy: 0.9973 - val_loss: 0.0562 - val_accuracy: 0.9864
Epoch 173/400
46/46 [==============================] - 0s 3ms/step - loss: 0.0023 - accuracy: 0.9986 - val_loss: 0.0547 - val_accuracy: 0.9864

46/46 [==============================] - 0s 3ms/step - loss: 0.0018 - accuracy: 0.9986 - val_loss: 0.0585 - val_accuracy: 0.9796
Epoch 225/400
46/46 [==============================] - 0s 2ms/step - loss: 0.0018 - accuracy: 0.9986 - val_loss: 0.0583 - val_accuracy: 0.9796
Epoch 226/400
46/46 [==============================] - 0s 3ms/step - loss: 0.0017 - accuracy: 0.9986 - val_loss: 0.0581 - val_accuracy: 0.9796
Epoch 227/400
46/46 [==============================] - 0s 3ms/step - loss: 0.0017 - accuracy: 0.9986 - val_loss: 0.0592 - val_accuracy: 0.9796
Epoch 228/400
46/46 [==============================] - 0s 2ms/step - loss: 0.0017 - accuracy: 0.9986 - val_loss: 0.0596 - val_accuracy: 0.9796
Epoch 229/400
46/46 [==============================] - 0s 2ms/step - loss: 0.0017 - accuracy: 0.9980 - val_loss: 0.0603 - val_accuracy: 0.9796
Epoch 230/400
46/46 [==============================] - 0s 3ms/step - loss: 0.0016 - accuracy: 0.9986 - val_loss: 0.0607 - val_accuracy: 0.9796
Epoch 231/400

46/46 [==============================] - 0s 3ms/step - loss: 0.0014 - accuracy: 0.9986 - val_loss: 0.1002 - val_accuracy: 0.9796
Epoch 282/400
46/46 [==============================] - 0s 3ms/step - loss: 0.0014 - accuracy: 0.9986 - val_loss: 0.1003 - val_accuracy: 0.9796
Epoch 283/400
46/46 [==============================] - 0s 3ms/step - loss: 0.0013 - accuracy: 0.9986 - val_loss: 0.0997 - val_accuracy: 0.9796
Epoch 284/400
46/46 [==============================] - 0s 3ms/step - loss: 0.0013 - accuracy: 0.9986 - val_loss: 0.0996 - val_accuracy: 0.9796
Epoch 285/400
46/46 [==============================] - 0s 3ms/step - loss: 0.0013 - accuracy: 0.9986 - val_loss: 0.1006 - val_accuracy: 0.9728
Epoch 286/400
46/46 [==============================] - 0s 3ms/step - loss: 0.0013 - accuracy: 0.9986 - val_loss: 0.0994 - val_accuracy: 0.9796
Epoch 287/400
46/46 [==============================] - 0s 3ms/step - loss: 0.0013 - accuracy: 0.9986 - val_loss: 0.0999 - val_accuracy: 0.9796
Epoch 288/400

46/46 [==============================] - 0s 2ms/step - loss: 0.0012 - accuracy: 0.9986 - val_loss: 0.1052 - val_accuracy: 0.9728
Epoch 339/400
46/46 [==============================] - 0s 2ms/step - loss: 0.0012 - accuracy: 0.9986 - val_loss: 0.1064 - val_accuracy: 0.9728
Epoch 340/400
46/46 [==============================] - 0s 2ms/step - loss: 0.0012 - accuracy: 0.9986 - val_loss: 0.1076 - val_accuracy: 0.9728
Epoch 341/400
46/46 [==============================] - 0s 3ms/step - loss: 0.0011 - accuracy: 0.9986 - val_loss: 0.1076 - val_accuracy: 0.9728
Epoch 342/400
46/46 [==============================] - 0s 3ms/step - loss: 0.0011 - accuracy: 0.9986 - val_loss: 0.1076 - val_accuracy: 0.9728
Epoch 343/400
46/46 [==============================] - 0s 3ms/step - loss: 0.0011 - accuracy: 0.9986 - val_loss: 0.1098 - val_accuracy: 0.9728
Epoch 344/400
46/46 [==============================] - 0s 2ms/step - loss: 0.0011 - accuracy: 0.9986 - val_loss: 0.1085 - val_accuracy: 0.9728
Epoch 345/400

46/46 [==============================] - 0s 2ms/step - loss: 0.0010 - accuracy: 0.9986 - val_loss: 0.0956 - val_accuracy: 0.9796
Epoch 396/400
46/46 [==============================] - 0s 3ms/step - loss: 0.0010 - accuracy: 0.9986 - val_loss: 0.0954 - val_accuracy: 0.9796
Epoch 397/400
46/46 [==============================] - 0s 3ms/step - loss: 0.0010 - accuracy: 0.9986 - val_loss: 0.0969 - val_accuracy: 0.9796
Epoch 398/400
46/46 [==============================] - 0s 3ms/step - loss: 0.0010 - accuracy: 0.9986 - val_loss: 0.0951 - val_accuracy: 0.9796
Epoch 399/400
46/46 [==============================] - 0s 3ms/step - loss: 0.0011 - accuracy: 0.9986 - val_loss: 0.0968 - val_accuracy: 0.9796
Epoch 400/400
46/46 [==============================] - 0s 3ms/step - loss: 0.0010 - accuracy: 0.9986 - val_loss: 0.0990 - val_accuracy: 0.9796


In [34]:
accurate_train_accuracy,classify_report= test_model(model,X_train,Y_train)

46/46 [==============================] - 0s 928us/step


In [35]:
accurate_train_accuracy

0.998639455782313

In [36]:
print(classify_report)

              precision    recall  f1-score   support

           0       1.00      1.00      1.00       272
           1       0.99      1.00      1.00       278
           2       1.00      0.99      1.00       376
           3       1.00      1.00      1.00       544

    accuracy                           1.00      1470
   macro avg       1.00      1.00      1.00      1470
weighted avg       1.00      1.00      1.00      1470



In [37]:
accurate_val_accuracy,classify_report= test_model(model,X_val,Y_val)

5/5 [==============================] - 0s 1ms/step


In [38]:
accurate_val_accuracy

0.9795918367346939

In [39]:
print(classify_report)

              precision    recall  f1-score   support

           0       1.00      0.95      0.97        20
           1       0.95      0.95      0.95        20
           2       0.95      1.00      0.97        36
           3       1.00      0.99      0.99        71

    accuracy                           0.98       147
   macro avg       0.97      0.97      0.97       147
weighted avg       0.98      0.98      0.98       147



In [40]:
accurate_test_accuracy,classify_report= test_model(model,X_test,Y_test)

5/5 [==============================] - 0s 1ms/step


In [41]:
accurate_test_accuracy

0.9932432432432432

In [42]:
print(classify_report)

              precision    recall  f1-score   support

           0       0.95      1.00      0.97        18
           1       1.00      0.95      0.97        20
           2       1.00      1.00      1.00        45
           3       1.00      1.00      1.00        65

    accuracy                           0.99       148
   macro avg       0.99      0.99      0.99       148
weighted avg       0.99      0.99      0.99       148



## Notes

- name            : first model 
- split           : 1/3 
- oversampling    : 0.5 
- val_acc         : 0.9836734693877551
- test_acc        : 0.983739837398374
- gamma           : 1 
---------------------------------------------------------------------------------------------
- name            : second model 
- split           : 1/3 
- oversampling    : 0.8
- val_acc         : 0.9877551020408163
- test_acc        : 0.9878048780487805
- gamma           : 1 

---------------------------------------------------------------------------------------------

- name            : forth model 
- split           : 1/5
- oversampling    : 1
- val_acc         : 0.9863945578231292
- test_acc        : 0.9932432432432432
- gamma           : 2

---------------------------------------------------------------------------------------------
- name            : fifth model 
- split           : 1/5
- oversampling    : 0.5
- val_acc         : 0.9863945578231292
- test_acc        : 0.9864864864864865
- gamma           : 2

---------------------------------------------------------------------------------------------

